File: src/main/java/com/akagiyui/drive/entity/Folder.java
package com.akagiyui.drive.entity;

import com.akagiyui.common.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.Data;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.hibernate.annotations.DynamicInsert;

import java.util.List;

/**
 * 文件夹 实体
 *
 * @author AkagiYui
 */
@Data
@ToString(exclude = {"parent", "user"})
@Accessors(chain = true)
@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"name", "parent_id", "user_id"})
})
@DynamicInsert
public class Folder extends BaseEntity {
    /**
     * 文件夹名
     */
    @Column(nullable = false)
    private String name;

    /**
     * 父文件夹
     */
    @JoinColumn(name = "parent_id")
    @ManyToOne(fetch = FetchType.EAGER)
    private Folder parent;

    /**
     * 所属用户
     */
    @JoinColumn(name = "user_id", nullable = false)
    @ManyToOne(fetch = FetchType.EAGER)
    private User user;

    /**
     * 是否在根目录下
     */
    public boolean inRoot() {
        return parent == null;
    }

    /**
     * 获取完整路径
     */
    public String getPath() {
        if (inRoot()) {
            return "/" + name;
        } else {
            return parent.getPath() + "/" + name;
        }
    }

    /**
     * 获取层级路径列表
     */
    public List<String> getPathList() {
        if (inRoot()) {
            return List.of(name);
        } else {
            List<String> pathList = parent.getPathList();
            pathList.add(name);
            return pathList;
        }
    }
}

File: app/src/main/kotlin/com/akagiyui/drive/entity/Folder.kt
package com.akagiyui.drive.entity

import com.akagiyui.common.entity.BaseEntity
import jakarta.persistence.*
import org.hibernate.annotations.DynamicInsert

/**
 * 文件夹 实体
 *
 * @author AkagiYui
 */
@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["name", "parent_id", "user_id"])])
@DynamicInsert
class Folder : BaseEntity() {
    /**
     * 文件夹名
     */
    @Column(nullable = false)
    lateinit var name: String

    /**
     * 父文件夹
     */
    @JoinColumn(name = "parent_id")
    @ManyToOne(fetch = FetchType.EAGER)
    var parent: Folder? = null

    /**
     * 所属用户
     */
    @JoinColumn(name = "user_id", nullable = false)
    @ManyToOne(fetch = FetchType.EAGER)
    lateinit var user: User

    /**
     * 是否在根目录下
     */
    val inRoot
        get() = parent == null

    /**
     * 获取完整路径
     */
    val path: String
        get() = if (inRoot) {
            "/$name"
        } else {
            "${parent!!.path}/$name"
        }

    /**
     * 获取层级路径列表
     */
    val pathList: MutableList<String?>
        get() = if (inRoot) {
            mutableListOf(name)
        } else {
            parent!!.pathList.apply {
                add(name)
            }
        }

}
